File: backend/app/api/v1/definitions.py
from fastapi import APIRouter, HTTPException, Query, status

from app.schemas.dictionary import WordDefinitionResponse, WordPronunciationResponse
from app.services.dictionary_service import dictionary_service

router = APIRouter()


@router.get("/{word}", response_model=WordDefinitionResponse)
async def get_word_definition(word: str):
    """Get definition of a word from external dictionary API"""

    try:
        definition = await dictionary_service.get_word_definition(word)
        return definition
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.get("/{word}/pronounce", response_model=WordPronunciationResponse)
async def get_word_pronunciation(
    word: str, voice: str = Query(default="us", pattern="^(us|uk|au)$")
):
    """Get pronunciation audio URL for a word"""

    try:
        pronunciation = await dictionary_service.get_word_pronunciation(word, voice)
        return pronunciation
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))

File: frontend/src/services/definitions.ts
import { fetchAPI } from './api';

export interface WordDefinitionItem {
  partOfSpeech: string;
  definition: string;
  example?: string;
}

export interface WordDefinition {
  word: string;
  definitions: WordDefinitionItem[];
  phonetic?: string;
  audioUrl?: string;
}

export interface PronunciationResponse {
  word: string;
  audioUrl: string;
  phonetic?: string;
}

// Get word definition
export async function getWordDefinition(word: string): Promise<WordDefinition> {
  return fetchAPI<WordDefinition>(`/definitions/${encodeURIComponent(word)}`);
}

// Get word pronunciation
export async function getWordPronunciation(
  word: string,
  voice: 'us' | 'uk' | 'au' = 'us'
): Promise<PronunciationResponse> {
  return fetchAPI<PronunciationResponse>(
    `/definitions/${encodeURIComponent(word)}/pronounce?voice=${voice}`
  );
}
